Introduce improved release workflow

tidoust opened the pull request:
This updates implements the release workflow described in:
https://github.com/w3c/webref/issues/145#issuecomment-805668484

As opposed to the current release workflow, this workflow will release the version that already appears in the `package.json` files, and not the bumped version. Hence the patch bump on these files in this update.

The pre-release job takes care of creating a pre-release PR whenever needed. The job updates the pre-release PR when the diff (or the version to release) changes.

The job no longer runs on a fixed schedule, and rather runs whenever a push is made that could create a diff with the current npm package. This may mean that there will often be an open pre-release PR to review. If CSS and IDL change too much, we could perhaps only assign reviewers to the pre-release PR only once per week, instead of immediately when the PR gets created.

The second job runs when a pre-release PR gets merged, and releases the package to npm. The content of the package is taken at the main branch commit on which the pre-release PR is based.

tidoust commented:
Tests fail due to IDL inconsistency, not because of the changes in this PR.

foolip commented on .github/workflows/release-package.yml:
```diff
@@ -0,0 +1,33 @@
+        node-version: 14.x
+
+    - name: Checkout webref
+      uses: actions/checkout@v2
```
Will this check out the branch of the PR that was closed? Or a merge commit of that and main, as usually happens with PRs?

tidoust replied:
That's unrelated to the PR. This will checkout the latest commit on the main branch. This checkout is only needed to be able to run the latest version of the `release-package.js` script.

tidoust replied:
Sorry, I just realized that my previous comment is totally wrong. The [checkout action](https://github.com/actions/checkout#checkout-v2) will checkout the ref/sha that triggered the workflow. In this case, this will be the [last merge commit on the PR branch](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request).

In turn, this means that possible fixes/changes made to the release script after a pre-release PR has been created would not be taken into account when the PR is merged. That seems counter-intuitive and not a good thing. I'll update the job to checkout the main branch instead.

foolip commented on .github/workflows/release-package.yml:
```diff
@@ -0,0 +1,33 @@
+  pull_requests:
+    branches:
+      - 'release-*'
+    types: [closed]
```
Will this trigger is the PR is closed without merging too?

tidoust replied:
Yes, hence the need for the `if: github.event.pull_request.merged == true` check 4 lines below. I could not find a direct way to trigger a worfklow on a merge action.

foolip commented on tools/prepare-release.js:
```diff
@@ -0,0 +1,305 @@
+  // The "diff" command exits with a non zero status code when there is a diff,
+  // which would throw an exception. Final "echo" command turns that status
+  // code to 0 to avoid the exception.
+  const installedIdl = path.join(tmpFolder, "node_modules", "@webref", type);
```
Might be CSS, not necessarily IDL?

foolip commented on tools/prepare-release.js:
```diff
@@ -0,0 +1,305 @@
+  // code to 0 to avoid the exception.
+  const installedIdl = path.join(tmpFolder, "node_modules", "@webref", type);
+  let diff = execSync(
+    `diff ${installedIdl} packages/${type} --ignore-trailing-space --exclude=package.json --unified=3 || echo -n`,
```
Will this list newly added files? I noticed the latest release has new IDL files which didn't seem to be listed in the prepared PR.

tidoust replied:
Looking at the most recent pre-release PR (https://github.com/w3c/webref/pull/208), I see that it did report on newly added files:
```
[...]
Only in webref/packages/idl: sanitizer-api.idl
Only in webref/packages/idl: scheduling-apis.idl
[...]
Only in webref/packages/idl: web-animations-2.idl
```

These lines are very easy to miss though, especially since they are not coloured as the rest of the diffs. I cannot find a diff option to make them stand out a bit more though :( For instance, it would be good if there was a way to add a newline between diff outputs

dontcallmedom replied:
we could grep "^Only in webref/packages/" into a separate output that would be presented separately from the rest of the diff?

foolip commented on tools/prepare-release.js:
```diff
@@ -0,0 +1,305 @@
+  }
+
+  // On Windows, the "echo -n" command actually outputs "-n"
+  // Also replace tmp folder name in diff to make it possible to compare diff
```
Have you looked for a way to invoke diff that omits the leading parts of the paths in the output? Since `patch -p3` can strip them I half suspect there's a way.

tidoust replied:
I [checked the `diff` manual](https://www.man7.org/linux/man-pages/man1/diff.1.html) and could not find one.

foolip commented on tools/prepare-release.js:
```diff
@@ -0,0 +1,305 @@
+
+  console.log("Look for a pending pre-release PR");
+  const searchResponse = await octokit.search.issuesAndPullRequests({
+    q: `repo:${owner}/${repo} type:pr state:open 📦 Release @webref/${type}@ in:title`
```
Shouldn't the title be quoted here?

And it might be easier to just use two unique branch name patterns and find the PR by branch name.

tidoust replied:
Both good points. I'll change that.

I was under the impression that quotes did not exist in GitHub's search API, but [quotes can be used](https://docs.github.com/en/github/searching-for-information-on-github/understanding-the-search-syntax#use-quotation-marks-for-queries-with-whitespace) indeed.

And I thought that branch name had to be the exact branch name to match, but that is also wrong, searching for `release-idl-` or `release-css-` works fine.

foolip commented on tools/prepare-release.js:
```diff
@@ -0,0 +1,305 @@
+  const latestCommitResponse = await octokit.repos.getCommit({
+    owner, repo, ref: "HEAD"
+  });
+  const latestCommitSha = latestCommitResponse?.data?.sha;
```
Using `?.` but still not checking if the result is null makes the code look undecided about the possibility of not finding the latest SHA.

foolip commented on tools/prepare-release.js:
```diff
@@ -0,0 +1,305 @@
+    console.log("== The end ==");
+  })
+  .catch(err => {
+    console.error(err);
```
Also `process.exit(1)`?

foolip commented on tools/prepare-release.js:
```diff
@@ -0,0 +1,305 @@
+  }
+})();
+if (!GH_TOKEN) {
+  console.error("Env variable GH_TOKEN must be set to some personal access token");
```
Or config.json

foolip commented on tools/prepare-release.js:
```diff
@@ -0,0 +1,305 @@
+
+✍ If any change needs to be made before release, **do not add a commit** to this pull request. Changes should rather be handled in a separate pull request and pushed to the default branch. You may leave this pull request open in the meantime, or close it. The pre-release job will automatically update this pull request or create a new one once the updates have made their way to the default branch.
+
+🛈 Merging this pull request will bump the patch part of the version number in \`packages/${type}/package.json\` so that a new release cycle may begin.
```
Maybe explain where the bump happens since it takes some pondering to understand the model?

tidoust replied:
I reformulated the text. I'm not fully convinced that the new text is clear, but when I try to add more text, I have the feeling that it actually becomes more confusing... Happy to take suggestions :)

In the end, what matters is to make it clear that the file changed by the PR is uninteresting for the purpose of issuing a new release. It is just there to reset the cycle after the package has been released.

foolip commented on tools/release-package.js:
```diff
@@ -0,0 +1,158 @@
+    console.log("== The end ==");
+  })
+  .catch(err => {
+    console.error(err);
```
Ditto

dontcallmedom commented:
> The job no longer runs on a fixed schedule, and rather runs whenever a push is made that could create a diff with the current npm package. This may mean that there will often be an open pre-release PR to review.

If I understand correctly, this means we will have pre-releases pull requests opened or updated 2 or 3 times a day outside of week-ends. What's the rationale for that approach vs scheduled pre-releases? Invalidating pull request reviews on updates (which I see marked as a TODO) might be quite important to avoid merging unreviewed changes.

tidoust commented:
> If I understand correctly, this means we will have pre-releases pull requests opened or updated 2 or 3 times a day outside of week-ends.

The code runs on week-ends too ;)

> What's the rationale for that approach vs scheduled pre-releases?

Less release code to maintain. The same job does the creation of the PR and the management of that PR.

> Invalidating pull request reviews on updates (which I see marked as a TODO) might be quite important to avoid merging unreviewed changes.

To avoid merging unreviewed changes, I guess I would propose another approach to start with: encourage reviewers to merge the release directly after review, instead of just issuing an ok without merging. Do we really need two or more reviews for now?

My worry is more that, if we are in a situation where there always is a pre-release PR that is pending review, then reviewers are going to ignore notification messages and no one is going to review and merge the pull request. It may make sense to only request a review only once per week.

tidoust commented:
> merging directly is OK, but I was worried about publishing too many package updates; but maybe that's not worth worrying about.

I updated the pre-release job not to add reviewers to start with, and added a job that runs weekly on Thursday to add reviewers. Goal is to strike a good balance between releasing too often and forgetting to release when needed. We may improve on that later on.

foolip commented:
I won't have time to further review this this week. Sorry for the pile of comments and running, feel free to ignore anything that didn't make sense.

foolip commented:
> I may be missing something but how would that simplify the workflow?

The main thing I was thinking was that both the before and after state would be an npm package. But since it ends up just comparing files on the filesystem, it's not really any simpler.

tidoust commented:
Discussed with @dontcallmedom. A new version of the IDL package has just been published. I will now merge this. First thing that should happen is that a new PR will be created to publish an updated CSS package. And then we'll see!
